staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go

```diff
@@ -46,6 +48,75 @@ const (
 	kmsPluginConnectionTimeout   = 3 * time.Second
 )
 
+// KMSPluginProbe checks health status of KMS-Plugins.
+type KMSPluginProbe struct {
+	// Name uniquely identifies probes. This value is taken from the name of the corresponding KMS Provider.
+	Name string
+	envelope.Service
+}
+
+// GetKMSPluginProbes extracts KMSPluginProbes from the EncryptionConfig.
+func GetKMSPluginProbes(filepath string) ([]*KMSPluginProbe, error) {
+	f, err := os.Open(filepath)
+	if err != nil {
+		return nil, fmt.Errorf("error opening encryption provider configuration file %q: %v", filepath, err)
+	}
+	defer f.Close()
+	return getKMSPluginProbes(f)
+}
+
+func getKMSPluginProbes(reader io.Reader) ([]*KMSPluginProbe, error) {
+	var result []*KMSPluginProbe
+
+	configFileContents, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return result, fmt.Errorf("could not read content of encryption provider configuration: %v", err)
+	}
+
+	config, err := loadConfig(configFileContents)
+	if err != nil {
+		return result, fmt.Errorf("error while parsing encrypiton provider configuration: %v", err)
+	}
+
+	for _, r := range config.Resources {
+		for _, p := range r.Providers {
+			if p.KMS != nil && p.KMS.Probe {
+				timeout := kmsPluginConnectionTimeout
+				if p.KMS.Timeout != nil {
+					if p.KMS.Timeout.Duration <= 0 {
+						return nil, fmt.Errorf("could not configure KMS-Plugin's probe %q, timeout should be a positive value", p.KMS.Name)
+					}
+					timeout = p.KMS.Timeout.Duration
+				}
+
+				s, err := envelope.NewGRPCService(p.KMS.Endpoint, timeout)
+				if err != nil {
+					return nil, fmt.Errorf("could not configure KMS-Plugin's probe %q, error: %v", p.KMS.Name, err)
+				}
+				result = append(result, &KMSPluginProbe{Name: p.KMS.Name, Service: s})
+			}
+		}
+	}
+
+	return result, nil
+}
+
+// Check encrypts and decrypts test data against KMS-Plugin's gRPC endpoint.
+func (h *KMSPluginProbe) Check() error {
+	p, err := h.Service.Encrypt([]byte("ping"))
```

Comment:
this connects the KMS directly to the apiserver /healthz endpoint (which is typically allowed to be triggered unauthenticated). do we want a 1:1 relationship or do we need to make sure the KMS isn't asked to encode/decode that frequently?

Reply:
I feel that rate-limiting should be the responsibility of the plugin (for example, Google's CloudKMS plugin implements client-side rate-limiting), so I think 1:1 is OK from kube-apiserver's point of view.

Reply:
would you expect the plugin to rate limit by failing or blocking requests? either way, since the plugin can't distinguish between this health check and encrypt/decrypt requests made in service of secret reads/writes, it doesn't seem great to allow pings of the /healthz endpoint to interfere to that degree with KMS operations needed to service the API

Reply:
You are right - the issue here is that KMS Plugin can't tell whether it is being health-checked or is servicing a valid request.

I would like to address this issue (on a separate PR) by adding support for a Health Checking Protocol for KMS-Plugin's gRPC (as was suggested by @awly  ).
See: https://github.com/grpc/grpc/blob/master/doc/health-checking.md
This way, kms-plugin can make more stringent rate-limiting decisions when receiving high volume of traffic on the Health endpoint.

From kube-apiserver's point of view, it could first try the Health endpoint and fall-back to what is currently implemented in this PR (encrypt/decrypt round-trip).

Let me know if it would make sense to add Health Checking to KMS-Plugin on this PR.

Reply:
a separate PR for a dedicated KMS health check is fine, but I would rate limit how frequently we hit the KMS encrypt/decrypt endpoints for health-checking purposes in this PR

Reply:
Added a rate-limiter - 1 healthz check per second.